### PR TITLE
refactor(xwayland): improve dnd and cleanup

### LINKED ISSUE #8964: [xwayland] drag&drop not emitting `drag_end`
### Already reported ? *

- [X] I have searched the existing open and closed issues.

### Regression?

Not a regression - it's bug regarding new feature

### System Info and Hyprland Version

Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887  (version: bump to 0.46.2).
Date: Thu Dec 19 19:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.5.1
 hyprlang 0.6.0
 hyprutils 0.3.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.1

flags set:
debug

System Information:
System name: Linux
Node name: marlo
Release: 6.12.8-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Thu, 02 Jan 2025 22:52:26 +0000

GPU information: 
03:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Navi 31 [Radeon RX 7900 XT/7900 XTX/7900 GRE/7900M] [1002:744c] (rev c8) (prog-if 00 [VGA controller])
19:00.0 VGA compatible controller [0300]: Advanced Micro Devices, Inc. [AMD/ATI] Raphael [1002:164e] (rev c1) (prog-if 00 [VGA controller])

os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo

plugins:

### Description

Drag&Dropping into an xwayland application (f.e. discord) does not emit a drag_end event. 

Tested with https://github.com/nik012003/ripdrag and https://github.com/abenz1267/walker.

I'm not sure if this only affects GTK4 apps or not....

### How to reproduce

Use ripdrag or walker to drag&drop something into discord (xwayland). Ripdrag won't close and walker gets stuck as well, because the drag_end event is not being triggered.

### Attach not paste

- [X] I understand that all text files must be *attached*, and not pasted directly. If not respected, this issue will likely get closed as spam

### Checklist of files to include below

- [X] Hyprland config - `hyprctl systeminfo -c` (always include)
- [ ] Crash report (always include in case of crash)
- [ ] Video (always include in case of a visual bug)
- [ ] Logs (might contain useful info such as errors)

### Additional info & File uploads

[hypr.txt](https://github.com/user-attachments/files/18310409/hypr.txt)

### LINKED ISSUE #8994: DND from wayland to xwayland window ignores XdndProxy property
### Already reported ? *

- [X] I have searched the existing open and closed issues.

### Regression?

Not a regression - it's an old bug

### System Info and Hyprland Version

<details> <summary>System/Version info</summary>

```
Hyprland 0.46.2 built from branch  at commit 0bd541f2fd902dbfa04c3ea2ccf679395e316887  (version: bump to 0.46.2).
Date: Thu Dec 19 19:26:47 2024
Tag: v0.46.2, commits: 5566
built against:
 aquamarine 0.5.1
 hyprlang 0.6.0
 hyprutils 0.3.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.1


flags set:
debug


System Information:
System name: Linux
Node name: archdome
Release: 6.12.4-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Mon, 09 Dec 2024 14:31:57 +0000


GPU information: 
02:00.0 VGA compatible controller [0300]: NVIDIA Corporation GA104 [GeForce RTX 3070] [10de:2484] (rev a1) (prog-if 00 [VGA controller])
NVRM version: NVIDIA UNIX x86_64 Kernel Module  565.77  Wed Nov 27 23:33:08 UTC 2024


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:


```

</details>

### Description

No DND events are delivered to xwayland subwindows that are referenced by a `XdndProxy` property on the toplevel window when the drag-source is a wayland window.  This type of window-setup is common in DAW plugin-UIs that want to accept audio-samples via DND for example.
Dragging from an xwayland window to such a target works.
I did a quick search through the codebase and didn't find any usage of the `XdndProxy` atom, so this is probably a straightforward fix.

### How to reproduce

1. Create a toplevel X window and a subwindow
2. Set `XdndAware` on the subwindow
3. Set `XdndProxy` on the toplevel window pointing to the subwindow
4. Set `XdndProxy` on the subwindow pointing to itself
5. Validate drag from xwayland source gets reccognized
6. Try dragging from wayland source

### Attach not paste

- [X] I understand that all text files must be *attached*, and not pasted directly. If not respected, this issue will likely get closed as spam

### Checklist of files to include below

- [X] Hyprland config - `hyprctl systeminfo -c` (always include)
- [ ] Crash report (always include in case of crash)
- [ ] Video (always include in case of a visual bug)
- [ ] Logs (might contain useful info such as errors)

### Additional info & File uploads

_No response_

### LINKED ISSUE #9154: drag & drop blocks input even after finshing
### Already reported ? *

- [x] I have searched the existing open and closed issues.

### Regression?

Probably not a regression / I don't remember it happening before

### System Info and Hyprland Version

<details> <summary>System/Version info</summary>

```
Hyprland 0.46.0 built from branch  at commit a3e7f21516f378e1928486a853881ee4941c1fdf  ().
Date: Fri Jan 24 14:07:16 2025
Tag: v0.46.0-178-ga3e7f215, commits: 5733
built against:
 aquamarine 0.7.1
 hyprlang 0.6.0
 hyprutils 0.4.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.1


no flags were set


System Information:
System name: Linux
Node name: Arch-Linux
Release: 6.12.10-zen1-1-zen
Version: #1 ZEN SMP PREEMPT_DYNAMIC Sat, 18 Jan 2025 02:26:52 +0000


GPU information: 
00:02.0 VGA compatible controller [0300]: Intel Corporation Raptor Lake-S GT1 [UHD Graphics 770] [8086:a780] (rev 04) (prog-if 00 [VGA controller])
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation AD103 [GeForce RTX 4080 SUPER] [10de:2702] (rev a1) (prog-if 00 [VGA controller])
NVRM version: NVIDIA UNIX x86_64 Kernel Module  565.77  Wed Nov 27 23:33:08 UTC 2024


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:


```

</details>

### Description

drag and drop will sometimes still block input after finishing

for example if i drag and drop from firefox download and go to other workspaces, it will not work and continue to think im dragging when im not

### How to reproduce

1. drag & drop (e.g. firefox, blobdrop etc)
2. maybe it will not stop even after u dragged

### Attach not paste

- [x] I understand that all text files must be *attached*, and not pasted directly. If not respected, this issue will likely get closed as spam

### Checklist of files to include below

- [ ] Hyprland config - `hyprctl systeminfo -c` (always include)
- [ ] Crash report (always include in case of crash)
- [ ] Video (always include in case of a visual bug)
- [ ] Logs (might contain useful info such as errors)

### Additional info & File uploads

_No response_

### LINKED ISSUE #9472: File drag and drop will hang mouse clicks
### Already reported ? *

- [x] I have searched the existing open and closed issues.

### Regression?

Probably not a regression / I don't remember it happening before

### System Info and Hyprland Version

<details> <summary>System/Version info</summary>

```
Hyprland 0.47.2 built from branch  at commit 882f7ad7d2bbfc7440d0ccaef93b1cdd78e8e3ff  (version: bump to 0.47.2).
Date: Sun Feb 2 00:47:17 2025
Tag: v0.47.2, commits: 5767
built against:
 aquamarine 0.7.2
 hyprlang 0.6.0
 hyprutils 0.5.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.2


no flags were set


System Information:
System name: Linux
Node name: ZIZIonArch
Release: 6.13.3-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Mon, 17 Feb 2025 17:42:11 +0000


GPU information: 
0000:00:02.0 VGA compatible controller [0300]: Intel Corporation Meteor Lake-P [Intel Arc Graphics] [8086:7d55] (rev 08) (prog-if 00 [VGA controller])
	DeviceName: VGA


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:
```

</details>

### Description

Drag & drop files into some Xwayland appllication causes hyprland to not handle mouse clicks

### How to reproduce

Steps to reproduce:

Drag a file from Nautilus / Dolphin to an app like QQ or WeChat (both are running under Xwayland and due to some search results I believe the bug is present on any Xwayland application, but I do not have any at hand). The file drop would not be successful, and until I kill the process of Nautilus / Dolphin, Hyprland cannot recieve and mouse clicks.

### Attach not paste

- [x] I understand that all text files must be *attached*, and not pasted directly. If not respected, this issue will likely get closed as spam

### Checklist of files to include below

- [x] Hyprland config - `hyprctl systeminfo -c` (always include)
- [ ] Crash report (always include in case of crash)
- [ ] Video (always include in case of a visual bug)
- [ ] Logs (might contain useful info such as errors)

### Additional info & File uploads

<details><summary>`hyprctl systeminfo -c`</summary>

```
Hyprland 0.47.2 built from branch  at commit 882f7ad7d2bbfc7440d0ccaef93b1cdd78e8e3ff  (version: bump to 0.47.2).
Date: Sun Feb 2 00:47:17 2025
Tag: v0.47.2, commits: 5767
built against:
 aquamarine 0.7.2
 hyprlang 0.6.0
 hyprutils 0.5.0
 hyprcursor 0.1.11
 hyprgraphics 0.1.2


no flags were set


System Information:
System name: Linux
Node name: ZIZIonArch
Release: 6.13.3-arch1-1
Version: #1 SMP PREEMPT_DYNAMIC Mon, 17 Feb 2025 17:42:11 +0000


GPU information: 
0000:00:02.0 VGA compatible controller [0300]: Intel Corporation Meteor Lake-P [Intel Arc Graphics] [8086:7d55] (rev 08) (prog-if 00 [VGA controller])
	DeviceName: VGA


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

======Config-Start======


Config File: /home/atzlt/.config/hypr/hyprland.conf: Read Succeeded
env = LIBVA_DRIVER_NAME,nvidia
env = __GLX_VENDOR_LIBRARY_NAME,nvidia
monitor = , 2880x1800@120, auto, 2

xwayland {
    force_zero_scaling = true
}

env = XCURSOR_SIZE,24
env = GDK_SCALE, 2

env = HYPRCURSOR_SIZE,24
env = HYPRCURSOR_THEME,rose-pine-hyprcursor

$terminal = kitty
$fileMan = nautilus
$runner = tofi-drun
$browser = firefox

input {
    touchpad {
        natural_scroll = true
        scroll_factor = 0.2
        disable_while_typing = false
    }
}

general {
    gaps_out = 10
    gaps_in = 3
    col.inactive_border = rgba(00000000)
    layout = dwindle
}

decoration {
    rounding = 5
    rounding_power = 4
    inactive_opacity = 0.9
}

animations {
    bezier = myBezier, 0.05, 0.9, 0.1, 1.05

    animation = windowsMove, 1, 3, myBezier
    animation = windowsIn, 1, 2, default, popin 90%
    animation = windowsOut, 1, 1, default, popin 95%
    animation = border, 1, 2, default
    animation = fade, 1, 2, default
    animation = workspaces, 1, 3, default
}

misc {
    vfr = true
}

workspace = r[1-3], monitor:0

# Layer Rules
layerrule = blur, waybar
layerrule = ignorezero, waybar
layerrule = blur, notifications
layerrule = ignorezero, notifications
layerrule = blur, launcher
layerrule = ignorezero, launcher

# Window Rules
# pinned window style
windowrulev2 = bordercolor rgb(598da8) rgb(598da8), pinned:1
# windowrulev2 = float,class:(.*)
windowrulev2 = float,class:(QQ) # QQ Popups
windowrulev2 = float,class:(wechat) # QQ Popups
windowrulev2 = float,title:(Volume Control) # Volume Control
windowrulev2 = float,title:(Fcitx Configuration) # Fcitx Config
windowrulev2 = float,class:(clash-verge)

# Window Rules

$mainMod = SUPER
bind = $mainMod, Q, exec, $terminal
bind = $mainMod SHIFT, Q, exec, warp
bind = $mainMod, M, exit
bind = $mainMod, E, exec, $fileMan
bind = $mainMod, R, exec, $runner | xargs hyprctl dispatch exec --
bind = $mainMod, T, togglefloating,
bind = $mainMod CONTROL, DELETE, killactive,
bind = $mainMod, Tab, cyclenext,
bind = $mainMod, O, fullscreen, 1
bind = $mainMod, X, pin,
bind = $mainMod, Z, swapnext,
bind = $mainMod, G, togglegroup
bind = $mainMod, S, togglesplit
bind = $mainMod, F, exec, $browser
bind = $mainMod CONTROL, V, exec, cliphist list | dmenu -b -fn "BlexMonoNerdFont-23" -l 10 | cliphist decode | wl-copy
bind = $mainMod, L, exec, pidof hyprlock || hyprlock
bind = $mainMod, C, togglespecialworkspace, minimized
bind = $mainMod CONTROL, C, movetoworkspace, special:minimized
bind = $mainMod CONTROL, C, setfloating

bind = $mainMod_CONTROL, L, resizeactive, 50 0
bind = $mainMod_CONTROL, H, resizeactive, -50 0
bind = $mainMod_CONTROL, J, resizeactive, 0 50
bind = $mainMod_CONTROL, K, resizeactive, 0 -50

bind = , XF86MonBrightnessUp, exec, brightnessctl s +5%
bind = , XF86MonBrightnessDown, exec, brightnessctl s 5%-
bind = , XF86AudioRaiseVolume, exec, pamixer -i 5 
bind = , XF86AudioLowerVolume, exec, pamixer -d 5 
bind = , XF86AudioMicMute, exec, pamixer --default-source -m
bind = , XF86AudioMute, exec, pamixer -t

bind = $mainMod SHIFT, S, exec, grim -g "$(slurp)" - | wl-copy
bind = $mainMod SHIFT CONTROL, S, exec, wf-recorder -g "$(slurp)" -f "/home/atzlt/Videos/ScreenRecords/recording-$(date +%y%m%d%H%M%S).mp4"
bind = $mainMod, escape , exec, kill "$(pidof wf-recorder)"
bind = $mainMod SHIFT, W, exec, grim -g "$(hyprctl -j activewindow | jq -r '"\(.at[0]),\(.at[1]) \(.size[0])x\(.size[1])"')" - | wl-copy
bind = ,print , exec, grim - | wl-copy

bind = $mainMod CONTROL, 1, movetoworkspace, 1
bind = $mainMod CONTROL, 2, movetoworkspace, 2
bind = $mainMod CONTROL, 3, movetoworkspace, 3
bind = $mainMod CONTROL, 4, movetoworkspace, 4
bind = $mainMod CONTROL, 5, movetoworkspace, 5
bind = $mainMod CONTROL, 6, movetoworkspace, 6
bind = $mainMod CONTROL, 7, movetoworkspace, 7
bind = $mainMod CONTROL, 8, movetoworkspace, 8
bind = $mainMod CONTROL, 9, movetoworkspace, 9 
bind = $mainMod CONTROL, 0, movetoworkspace, 10

bind = $mainMod CONTROL SHIFT, 1, movetoworkspacesilent, 1
bind = $mainMod CONTROL SHIFT, 2, movetoworkspacesilent, 2
bind = $mainMod CONTROL SHIFT, 3, movetoworkspacesilent, 3
bind = $mainMod CONTROL SHIFT, 4, movetoworkspacesilent, 4
bind = $mainMod CONTROL SHIFT, 5, movetoworkspacesilent, 5
bind = $mainMod CONTROL SHIFT, 6, movetoworkspacesilent, 6
bind = $mainMod CONTROL SHIFT, 7, movetoworkspacesilent, 7
bind = $mainMod CONTROL SHIFT, 8, movetoworkspacesilent, 8
bind = $mainMod CONTROL SHIFT, 9, movetoworkspacesilent, 9 
bind = $mainMod CONTROL SHIFT, 0, movetoworkspacesilent, 10

bind = $mainMod, 1, workspace, 1
bind = $mainMod, 2, workspace, 2
bind = $mainMod, 3, workspace, 3
bind = $mainMod, 4, workspace, 4
bind = $mainMod, 5, workspace, 5
bind = $mainMod, 6, workspace, 6
bind = $mainMod, 7, workspace, 7
bind = $mainMod, 8, workspace, 8
bind = $mainMod, 9, workspace, 9 
bind = $mainMod, 0, workspace, 10

bind = $mainMod, left, movefocus, l
bind = $mainMod, right, movefocus, r
bind = $mainMod, up, movefocus, u
bind = $mainMod, down, movefocus, d

bind = $mainMod CONTROL, left, movewindow, l
bind = $mainMod CONTROL, right, movewindow, r
bind = $mainMod CONTROL, up, movewindow, u
bind = $mainMod CONTROL, down, movewindow, d

bindm = $mainMod, mouse:272, movewindow
bindm = $mainMod, mouse:273, resizewindow

bindl = , switch:on:Lid Switch, exec, pidof hyprlock || hyprlock, hyprctl dispatch dpms off
bindl = , switch:off:Lid Switch, exec, hyprctl dispatch dpms on

plugin {
    hyprwinwrap {
        class = classwallpaper
    }
}

exec-once = hyprpaper
exec-once = pulseaudio
exec-once = gsettings set org.gnome.desktop.interface cursor-theme BreezeX-RosePine-Linux
exec-once = gsettings set org.gnome.desktop.interface cursor-size 24
exec-once = wl-paste --watch cliphist store
exec-once = wl-clip-persist --clipboard regular
exec-once = waybar
exec-once = libinput-gestures-setup start
exec-once = hypridle
exec-once = mako
exec-once = gsettings set com.github.stunkymonkey.nautilus-open-any-terminal terminal kitty

exec-once = clash-verge

# Why doesn't it work?
# exec-once = ~/Projects/wallpaper/bin/python ~/Projects/wallpaper/classwallpaper

======Config-End========

```

</details>

### PR DESCRIPTION
**Pull request by nnyyxxxx**
fixes #9154 
fixes #8964 
fixes #8994 
fixes #9472

### CODE REVIEW

**Comment by nnyyxxxx** [comment]
some edge cases

**Comment by nnyyxxxx** [comment]
g2g

**Comment by vaxerski** [comment]
christ, I'll have to test this myself a bit because in the dnd code when you touch something then some fucking app implodes

**Comment by nnyyxxxx** [comment]
> christ, I'll have to test this myself a bit because in the dnd code when you touch something then some fucking app implodes

should be good lmk how that goes

**Comment by nnyyxxxx** [comment]
did you find any bugs with this? @vaxerski or nah

**Comment by vaxerski** [comment]
I should make a rule that every time you ping me I delay my review by 2 days

**Comment by vaxerski** [comment]
dnd dolphin -> kitty = broken by this
dnd dolphin -> chromium tabs = broken by this

took me just 2 minutes

**Comment by nnyyxxxx** [comment]
atleast its better than before lol

**Comment by nnyyxxxx** [comment]
can you explain what u mean by broken lol

**Comment by nnyyxxxx** [comment]
nvm i see now

**Comment by nnyyxxxx** [comment]
broke more shit

**Comment by nnyyxxxx** [comment]
kitty officially fixed

**Comment by nnyyxxxx** [comment]
image

**Comment by nnyyxxxx** [comment]
idk what is up with chromium tbh

**Comment by nnyyxxxx** [comment]
probably not an issue here

**Comment by nnyyxxxx** [comment]
cuz i can perfectly drop into any other xwayland app

**Comment by nnyyxxxx** [comment]
>chromium tabs = broken by this (in the middle is fine, onto the tab bar to open as new tab is broken)

cant drop into middle before [ba57bc0](https://github.com/hyprwm/Hyprland/pull/9405/commits/ba57bc0e5d669ed079ee6e245744edb4e4246a79)

**Comment by nnyyxxxx** [comment]
can you take another look, and see if anything else needs fixing other than chromium?

**Comment by vaxerski** [comment]
yea, sec

**Comment by vaxerski** [comment]
both issues look fixed for me

**Comment by nnyyxxxx** [comment]
okay ill undraft then

**Comment by nnyyxxxx** [comment]
a

**Comment by nnyyxxxx** [comment]
im having a stroke

**Comment by nnyyxxxx** [comment]
review when

**Comment by vaxerski** [comment]
when I have some time to test this thoroughly

**Comment by nnyyxxxx** [comment]
wouldnt you have to test thoroughly after i make changes according to your reviews as well? when you do review of course?

whats the point of reviewing after you test? wouldnt it be better to review now so you dont have to test for a second time after i make changes according to your reviews?

**Comment by nnyyxxxx** [comment]
going based off of your message ofc

**Comment by vaxerski** [comment]
I did skim through it and didnt find anything immediately wrong

**Comment by nnyyxxxx** [comment]
okay

**Comment by Zebra2711** [comment]
dnd from wayland to xcb works fine, but from xcb to wayland, it simply doesn't work.

**Comment by nnyyxxxx** [comment]
if it didnt work before i cant be asked to make it work now

**Comment by nnyyxxxx** [comment]
ive already spent like 10 hours trying to make this shiz work.

**Comment by Zebra2711** [comment]
dnd from dolphin (wayland) to dolphin(xcb) doesn't work either.

```
kf.coreaddons: Failed to retrieve files from portal: QDBusError("org.freedesktop.DBus.Error.AccessDenied", "Invalid transfer")
```

**Comment by nnyyxxxx** [comment]
skill issue

**Comment by Zebra2711** [comment]
> skill issue

Sorry, i restarted dbus, so it works for now :)

**Comment by nnyyxxxx** [comment]
> Sorry, i restarted dbus, so it works for now :)

image

**Comment by nnyyxxxx** [comment]
> dnd from wayland to xcb works fine, but from xcb to wayland, it simply doesn't work.

works for me

**Comment by nnyyxxxx** [comment]
my code is impeccable

**Comment by nnyyxxxx** [comment]
my messages contain too much aura, gotta turn it down a bit

**Comment by nnyyxxxx** [comment]
cooking something up right neow

**Comment by nnyyxxxx** [comment]
7a6fde8414ae200759ac41569f189ac5e1ba0fbb this commit is too good

**Comment by nnyyxxxx** [comment]
nvm i lost interest already lol way too much work

File: src/xwayland/Dnd.hpp [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
 #include "../protocols/types/DataDevice.hpp"
+#include "../managers/SeatManager.hpp"
+#include "../managers/input/InputManager.hpp"
 #include <wayland-server-protocol.h>
 #include <hyprutils/os/FileDescriptor.hpp>
+#include <xcb/xcb.h>
 
 #define XDND_VERSION 5
 
+#define XCB_PROPERTY_FORMAT_32BIT 32
```
Comment (vaxerski) [inline src/xwayland/Dnd.hpp:12]: this is not an XCB define, it shouldnt have an XCB_ prefix.

**Review by vaxerski — COMMENTED** [review]
I did some testing and only found one minor bug, up to you if you want to fix it or not,

open up chromium in xwayland and dolphin in wayland
drag a file into chromium
dolphin is frozen (still prolly thinks dnd is going on) until you click on it

File: src/xwayland/Dnd.hpp (hunk 1 above)
Comment (nnyyxxxx) [reply src/xwayland/Dnd.hpp:12]: lol ok

**Comment by nnyyxxxx** [comment]
> I did some testing and only found one minor bug, up to you if you want to fix it or not,
> 
> open up chromium in xwayland and dolphin in wayland drag a file into chromium dolphin is frozen (still prolly thinks dnd is going on) until you click on it

lemme try

**Comment by nnyyxxxx** [comment]
dolphin doesnt freezer for me but it also doesnt drop into chromium

**Comment by nnyyxxxx** [comment]
dolphin is weird, probably not an issue with this

**Comment by nnyyxxxx** [comment]
okay... this is ready to merge

**Comment by nnyyxxxx** [comment]
https://www.youtube.com/watch?v=a9mBlX5uPbA

watch this before you merge :p

**Comment by nnyyxxxx** [comment]
whens the next release? we cooked up some good shiz in the past couple of weeks.....

File: src/xwayland/Dnd.hpp [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
+#include "../managers/input/InputManager.hpp"
 #include <wayland-server-protocol.h>
 #include <hyprutils/os/FileDescriptor.hpp>
+#include <xcb/xcb.h>
 
 #define XDND_VERSION 5
 
+#define PROPERTY_FORMAT_32BIT 32
+#define XCB_PROPERTY_LENGTH   1
+#define XCB_PROPERTY_OFFSET   0
```
Comment (vaxerski) [inline src/xwayland/Dnd.hpp:14]: these two too my friend

Comment (nnyyxxxx) [reply src/xwayland/Dnd.hpp]: ok....... i dont think it matters though

**Comment by nnyyxxxx** [comment]
anything else or is this good

**Comment by nnyyxxxx** [comment]
when is this getting merged?

**Comment by Zebra2711** [comment]
> dolphin is frozen

then, something weird happens [#9472](https://github.com/hyprwm/Hyprland/issues/9472)
same issue when dnd video from dolphin to the mpv window [#15939](https://github.com/mpv-player/mpv/issues/15939). I think this bug comes from hypland dnd too. Now, it is fixed with [this patch](https://patch-diff.githubusercontent.com/raw/mpv-player/mpv/pull/15940.patch) from mpv, but I don't think it will not happen in the future with another app.

**Comment by nnyyxxxx** [comment]
> then, something weird happens https://github.com/hyprwm/Hyprland/issues/9472

that is unrelated to

> dolphin is frozen

infact dolphin isnt even frozen, it just doesnt realize the drop has ended

**Comment by nnyyxxxx** [comment]
every other file manager i tested doesnt have this behavior

> infact dolphin isnt even frozen, it just doesnt realize the drop has ended

**Comment by nnyyxxxx** [comment]
tbh i dont even know why i put so much effort into this pull request, i dont even use a file manager for file management :sob:

**Comment by Zebra2711** [comment]
> > then, something weird happens #9472
> 
> that is unrelated to

No, i mean the same thing happens here is hyprland to not handle mouse clicks

**Comment by nnyyxxxx** [comment]
no... that is not the same issue... #9472 is related to input freezing on -git / 0.47.2

**Comment by nnyyxxxx** [comment]
dolphin doesn't freeze input in this mr after a dnd

**Comment by Zebra2711** [comment]
> dolphin doesn't freeze input in this mr after a dnd

yes, it cannot handle mouse clicks for any app on the screen. But the keyboard works fine

**Comment by nnyyxxxx** [comment]
> yes, it cannot handle mouse clicks for any app on the screen. But the keyboard works fine

???

before or after this mr

**Comment by nnyyxxxx** [comment]
bro is just yapping now...

**Comment by Zebra2711** [comment]
After

**Comment by nnyyxxxx** [comment]
cap

**Comment by nnyyxxxx** [comment]
me & @vaxerski already tested this... multiple times....

https://github.com/hyprwm/Hyprland/pull/9405#pullrequestreview-2634155241
https://github.com/hyprwm/Hyprland/pull/9405#issuecomment-2660936147

**Comment by Zebra2711** [comment]
> issue when dnd video from dolphin to the mpv window [#15939](https://github.com/mpv-player/mpv/issues/15939).

@nnyyxxxx have you tested this?

**Comment by nnyyxxxx** [comment]
> It's an event ordering issue on our part.

**Comment by nnyyxxxx** [comment]
https://github.com/mpv-player/mpv/pull/15940

**Comment by Zebra2711** [comment]
> I think this bug comes from hypland dnd too. Now, it is fixed with [this patch](https://patch-diff.githubusercontent.com/raw/mpv-player/mpv/pull/15940.patch) from mpv, but I don't think it will not happen in the future with another app.

**Comment by Zebra2711** [comment]
Why are you so toxic? Do you know how to read, or do you type faster than you think? If you want to be clear, just say, "That is not an issue with hyprland." So, should I create an issue for something other than hyprland instead? That's fine—I will create issues for dolphin, mpv, etc. But what I want to mention here is that I believe this issue comes from both sides. And one last thing: is your code impeccable?

**Comment by nnyyxxxx** [comment]
what :skull:

**Comment by nnyyxxxx** [comment]
https://www.youtube.com/watch?v=UnktCDi-BVs

**Comment by vaxerski** [comment]
@nnyyxxxx I do realize you are trying to contribute code, but you are repeatedly trying my patience. If you keep being this abrasive, you will be banned.

**Comment by nnyyxxxx** [comment]
my bad

**Comment by vaxerski** [comment]
can't dnd from dolphin to any xwayland app with this MR.

**Comment by nnyyxxxx** [comment]
i didnt even change anything

**Comment by vaxerski** [comment]
hm, right, it doesn't work on main either.

**Comment by nnyyxxxx** [comment]
try rebooting :skull:

**Comment by nnyyxxxx** [comment]
this pr works fine for me

**Comment by vaxerski** [comment]
9228116c9aa682477aae1a52d2d7d023bb075352 regressor

**Comment by nnyyxxxx** [comment]
??

**Comment by nnyyxxxx** [comment]
that doesnt even touch dnd

**Comment by vaxerski** [comment]
of course it does :)

**Comment by nnyyxxxx** [comment]
https://github.com/user-attachments/assets/927a58f7-7046-4545-befb-2df2b7b7616c

this is on latest -git

**Comment by vaxerski** [comment]
hm, okay, nevermind. It's just really dicey. It doesn't work with _some_ filetypes. Maybe some mime weirdness.

**Comment by vaxerski** [comment]
the dolphin bug is still there but it's way better than main, so I'm fine with this.

**Review by vaxerski — APPROVED** [review]

**Comment by vaxerski** [comment]
actually, sec.

**Comment by vaxerski** [comment]
there we go

**Review by vaxerski — APPROVED** [review]

**Comment by nnyyxxxx** [comment]
:skull:

**Comment by nnyyxxxx** [comment]
> but it's way better than main, so I'm fine with this.

of course

**Comment by johnhoh** [comment]
Dolphin dragging files to Davinci resolve(xwayland) still doesn't work.
